File: services/finances-frontend/src/utils/cashflow.ts
import type { Transaction, BankAccount } from '@/types/finances';

export interface CashflowSummary {
  income: number; // real income (excludes settlements & exchange accounts)
  incomeCount: number;
  incomeYield: number; // rendimentos (juros/yield)
  incomeYieldCount: number;
  incomeOther: number; // income that is not yield
  incomeOtherCount: number;
  expense: number; // real expenses (excludes settlements & exchange accounts)
  expenseCount: number;
}

/** Yield/interest income ("rendimento da conta", "Rendimento CDB", ...). */
export const isYield = (tx: Transaction): boolean => /rendimento/i.test(tx.description);

/**
 * Internal settlements that are NOT real income/expense and must be excluded from the
 * monthly cashflow:
 *  - credit-card invoice payments. The Pay flow records them as an INCOME credit on the
 *    card side AND as an EXPENSE "Pagamento fatura ..." on the funding account. Counting
 *    either as income/expense is wrong (the card purchases are already the real expense).
 */
export const isSettlement = (tx: Transaction, account?: BankAccount): boolean => {
  if (tx.type === 'INCOME' && account?.type === 'CREDIT_CARD') return true;
  if (/pagamento\s+fatura/i.test(tx.description)) return true;
  return false;
};

/** Summarize confirmed cashflow, separating yield from other income and dropping settlements. */
export const summarizeCashflow = (
  transactions: Transaction[],
  accounts: BankAccount[],
): CashflowSummary => {
  const accountById = new Map(accounts.map((a) => [a.id, a]));
  const exchangeIds = new Set(
    accounts.filter((a) => a.type === 'EXCHANGE').map((a) => a.id),
  );

  let income = 0;
  let incomeCount = 0;
  let incomeYield = 0;
  let incomeYieldCount = 0;
  let expense = 0;
  let expenseCount = 0;

  for (const tx of transactions) {
    if (tx.status !== 'CONFIRMED') continue;
    if (exchangeIds.has(tx.bankAccountId)) continue;
    if (isSettlement(tx, accountById.get(tx.bankAccountId))) continue;

    if (tx.type === 'INCOME') {
      income += tx.amount;
      incomeCount += 1;
      if (isYield(tx)) {
        incomeYield += tx.amount;
        incomeYieldCount += 1;
      }
    } else if (tx.type === 'EXPENSE') {
      expense += tx.amount;
      expenseCount += 1;
    }
  }

  return {
    income,
    incomeCount,
    incomeYield,
    incomeYieldCount,
    incomeOther: income - incomeYield,
    incomeOtherCount: incomeCount - incomeYieldCount,
    expense,
    expenseCount,
  };
};

File: services/calendar-finances/internal/domain/transaction/cashflow.go
package transaction

import (
	"strings"

	"github.com/brunovieira/calendar-finances/internal/domain/bankaccount"
	"github.com/brunovieira/calendar-finances/internal/domain/category"
)

// maxCategoryDepth stops a malformed parent chain from looping forever.
const maxCategoryDepth = 16

// IsSettlement reports whether the transaction only moves money around instead of
// earning or spending it, and therefore must stay out of a cashflow view.
//
// Three shapes qualify, all recognised structurally rather than by reading the
// description — a description is free text a person types, and a rule that depends on
// the wording silently breaks the month when someone writes it differently:
//
//   - the INCOME credit that lands on a credit card when its bill is paid. The
//     purchases on that card already were the expense.
//   - a TRANSFER into another account of the same owner — paying a card bill from
//     checking, funding an investment. Nothing was earned or spent.
//   - an expense that pays an invoice. Bills entered by hand carry the invoice link
//     but no destination, and they are the majority of the payments in this database.
//
// An EXPENSE carrying a destination is NOT a settlement: that is the source leg of a
// cross-profile transfer, where money genuinely left this profile.
//
// `account` is where the transaction sits and `destination` is the other end of a
// transfer, if any. A caller that cannot resolve them passes nil, and the transaction
// is treated as real — better to count a settlement than to drop real money.
func (t *Transaction) IsSettlement(account, destination *bankaccount.BankAccount) bool {
	if account != nil && account.IsCreditCard() && t.Type == TypeIncome {
		return true
	}
	if t.Type == TypeExpense && t.InvoiceID != nil && (account == nil || !account.IsCreditCard()) {
		return true
	}
	return t.Type == TypeTransfer && destination != nil
}

// IsYieldIn reports whether income came from the balance producing it — interest, a
// dividend, a fund's daily yield — rather than from work or a sale.
//
// The answer is the category's DRE bucket, inherited from ancestors exactly as the
// financial report resolves it: a subcategory "Rendimento CDB" under a FINANCIAL
// "Rendimentos" is yield, and reading only the leaf would disagree with the report on
// the same data. `categories` indexes the profile's categories by id so the chain can
// be walked without a repository call per transaction.
func (t *Transaction) IsYieldIn(cat *category.Category, categories map[string]*category.Category) bool {
	if t.Type != TypeIncome {
		return false
	}

	switch bucket := dreBucketOf(cat, categories); {
	case bucket == category.DREFinancial:
		return true
	case bucket != "":
		// A classified branch is the accounting answer, whatever the wording says.
		return false
	}

	// Nothing in the tree says anything. Almost nothing is classified yet — including
	// the personal "Rendimentos" that receives the daily account interest and every
	// dividend — so trusting the classification alone would silently report zero yield
	// for a whole profile. The wording decides until the categories are filled in, and
	// classifying one is what turns this off for it.
	return looksLikeYield(t.Description)
}

func looksLikeYield(description string) bool {
	return strings.Contains(strings.ToLower(description), "rendiment")
}

// dreBucketOf walks up to the nearest ancestor carrying a classification.
func dreBucketOf(cat *category.Category, categories map[string]*category.Category) category.ClassificationDRE {
	for depth := 0; cat != nil && depth < maxCategoryDepth; depth++ {
		if cat.ClassificationDRE != nil && *cat.ClassificationDRE != "" {
			return *cat.ClassificationDRE
		}
		if cat.ParentID == nil {
			return ""
		}
		cat = categories[*cat.ParentID]
	}
	return ""
}
